File: Data Structures & Algorithms/longest-consecutive-sequence/submission-0.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // step 1: Put everything into a hash set for O(1) lookups
        // This also handles duplicates automatically
        unordered_set<int> numSet(nums.begin(), nums.end());
        int longestStreak = 0;
        for(int num : numSet){
            // step 2: Check if 'num' is the start of a sequence
            // We only start counting if (num - 1) doesn't exist
            if(numSet.find(num - 1) == numSet.end()){
                int currentNum = num;
                int currentStreak = 1;

                // step 3 : keep checking for n + 1
                while (numSet.find(currentNum + 1) != numSet.end()){
                    currentNum += 1;
                    currentStreak += 1;
                }
                longestStreak = max(longestStreak, currentStreak);
            }
        
        }
        return longestStreak;
    }

};

File: Data Structures & Algorithms/house-robber-ii/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        def rob_range(start, end):
            prev2 = prev1 = 0
            for i in range(start, end + 1):
                current = max(prev1, nums[i] + prev2)
                prev2 = prev1
                prev1 = current
            return prev1
        
        return max(rob_range(0, len(nums)-2), rob_range(1, len(nums) - 1))